src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -227,6 +237,61 @@ async def mypy_typecheck_partition(
         requirements_venv_pex_request, file_list_digest_request
     )
 
+    py_version = config_file.python_version_to_autoset(
+        partition.interpreter_constraints, python_setup.interpreter_versions_universe
+    )
+    named_cache_dir = f".cache/mypy_cache/{sha256(build_root.path.encode()).hexdigest()}"
+    run_cache_dir = ".tmp_cache/mypy_cache"
+    argv = await _generate_argv(
+        mypy,
+        pex=mypy_pex,
+        venv_python=requirements_venv_pex.python.argv0,
+        cache_dir=run_cache_dir,
+        file_list_path=file_list_path,
+        python_version=py_version,
+    )
+
+    script_runner_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "__mypy_runner.sh",
+                    dedent(
+                        f"""\
+                            # We want to leverage the MyPy cache for fast incremental runs of MyPy.
+                            # Pants exposes "append_only_caches" we can leverage, but with the caveat
+                            # that it requires either only appending files, or multiprocess-safe access.
+                            #
+                            # MyPy guaruntees neither, but there's workarounds!
+                            #
+                            # By default, MyPy uses 2 cache files persource file, which introduces a
+                            # whole slew of race conditions. To workaround this we enable MyPy's SQLite cache,
+                            # which enables us to run MyPy on a temporary copy of the cache, then mv
+                            # the modified cache back to the real location. This is multiprocess-safe
+                            # as mv on the same filesystem is an atomic "rename", and any processes
+                            # copying the "old" file will still have valid file descriptors for the
+                            # "old" file.
+                            #
+                            # (MyPy uses the cache as a drop-in replacement for the dual-files solution,
```

Comment:
Great. So as to the other meaty 1/2 of things, why does cp/mv work at all or at least what is the proof? It would be good to mention that since it's not at all racing cp/mv db over-writes makes enough forward progress to be useful.

Reply:
It look like you amended this comment and added a middle paragraph addressing this: https://github.com/pantsbuild/pants/pull/16290#issuecomment-1195380807

Reply:
updated

Reply:
Thanks. That now seems to address the entirety of the tricks being used and why. I'm generally uncomfortable with claims of rare, mostly, etc - but you have confirmed things are actually a good deal faster with these tricks, which is what really matters.

I'll let the requested reviewers review the rest, I just wanted to make sure I understood exactly why this all worked. Thanks for bearing with me.

Reply:
I'm pleased with the resulting wording, so thanks for helping me along as well.